### DISCUSSION TITLE
configure: error: cannot run C compiled programs.

### DISCUSSION THREAD

**Discussion by banjo09** [rbenv/ruby-build#2266] [Build failures] [answered]
Command used: rbenv install 3.1.4

Output:

```
BUILD FAILED (macOS 14.0 using ruby-build 20230919)

Inspect or clean up the working tree at /var/folders/50/9023n_yn3y37_37kshp5psfh0000gn/T/ruby-build.20231010193520.25121.9KzEp1
Results logged to /var/folders/50/9023n_yn3y37_37kshp5psfh0000gn/T/ruby-build.20231010193520.25121.log

Last 10 log lines:
checking for llvm-strip... no
checking for gcc... clang
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables... 
checking whether we are cross compiling... configure: error: in `/var/folders/50/9023n_yn3y37_37kshp5psfh0000gn/T/ruby-build.20231010193520.25121.9KzEp1/ruby-3.1.4':
configure: error: cannot run C compiled programs.
If you meant to cross compile, use `--host'.
See `config.log' for more details
make: *** No targets specified and no makefile found.  Stop.
```

**Comment by oleksiivinogradov** [comment]
same here

**Comment by parkkh220706** [comment]
same problem here.. cannot install

**Comment by mislav** [comment, marked as answer]
Has anyone affected tried to install Xcode Command Line Tools with `xcode-select --install` or, if already installed, make sure they're upgraded through System Update?

**Reply by parkkh220706** [reply]
After System Update Xcode Command Line, rbenv install works properly. Thanks a lot.

**Reply by parkkh220706** [reply]
Have a question: Is Ruby 2.4.1 incompatible with macOS 14 Sonoma? 
Because when I run rbenv install 2.4.1, I get a 'BUILD FAILED (macOS 14.0 using ruby-build 20230919)' error. 
*ruby 3.1.4 installed well.

**Reply by banjo09** [reply]
I updated it through both. Then settled with updating it through xcode. 

it works fine now. Thanks

**Reply by mislav** [reply]
@parkkh220706 Yes, I would expect that Ruby 2.4 is not compatible with newer macOS releases. Official support for Ruby 2.4 ended in March 2020, meaning that all platform-specific bugs discovered since then (and macOS releases are known to frequently break things) would not get a fix.

Generally, you should only be installing Ruby 3+ versions on macOS Sonoma (all others are officially unsupported). Luckily, if you need Ruby 2.6, it comes built into Sonoma as `/usr/bin/ruby`.

**Reply by monfresh** [reply]
@parkkh220706 To clarify, Ruby 2.4 is in fact compatible with macOS Sonoma. However, it requires using Rosetta on an Apple Silicon Mac. Here is what I usually recommend:

- If you know what you're doing, try using Rosetta (it can be tricky to set up a proper dev environment that will let you easily switch between Rosetta mode and native Apple Silicon mode)
- Update your project to a newer Ruby version, at least 2.6.10, which doesn't require Rosetta. If you don't know how, here's my [step-by-step guide that can help you update the Ruby version in your project](https://www.rubyonmac.dev/how-to-upgrade-the-ruby-version-in-your-project)
- If updating the Ruby version is not an option at the moment, or if you need to be able to run the project with Ruby 2.4 in order to update it to a newer Ruby version, you can buy the Ultimate version of my [Ruby on Mac](https://www.rubyonmac.dev/pricing) product, which lets you easily install older Ruby versions with Rosetta and newer ones in native mode.

Whether you choose to install Ruby 2.4 by yourself or with Ruby on Mac, I still recommend updating your project to 2.4.10 using the instructions in my article above. There's usually no good reason to use a patch version (the 3rd digit) that's lower than the latest available one. [endoflife.date](https://endoflife.date/ruby) is a good site to see the latest versions of each release.

**Reply by parkkh220706** [reply]
@monfresh @mislav Thank you for kind responses. I appreciate it!

**Comment by iampato** [comment]
https://stackoverflow.com/questions/78548511/macos-sonoma-configure-error-cannot-run-c-compiled-programs

worked for me
